include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_gpu_submatrix.hpp

```diff
@@ -0,0 +1,365 @@
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE.txt for terms of usage.
+//  See CITATION.md for citation guidelines, if DCA++ is used for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch).
+
+// This class organizes the MC walker in the CT-INT QMC using a GPU accelerator.
+
+#ifndef DCA_PHYS_DCA_STEP_CLUSTER_SOLVER_CTINT_WALKER_CTINT_WALKER_GPU_SUBMATRIX_HPP
+#define DCA_PHYS_DCA_STEP_CLUSTER_SOLVER_CTINT_WALKER_CTINT_WALKER_GPU_SUBMATRIX_HPP
+
+#ifndef DCA_HAVE_CUDA
+#error "This file requires CUDA support."
+#endif
+
+#include "dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu_submatrix.hpp"
+
+#include <cassert>
+#include <stdexcept>
+#include <vector>
+
+#include "dca/linalg/util/cuda_event.hpp"
+#include "dca/phys/dca_step/cluster_solver/ctint/structs/device_configuration_manager.hpp"
+#include "dca/phys/dca_step/cluster_solver/ctint/walker/tools/d_matrix_builder_gpu.hpp"
+#include "dca/phys/dca_step/cluster_solver/ctint/walker/kernels_interface.hpp"
+
+namespace dca {
+namespace phys {
+namespace solver {
+namespace ctint {
+
+template <class Parameters, typename Real>
+class CtintWalkerSubmatrix<linalg::GPU, Parameters, Real>
+    : public CtintWalkerSubmatrix<linalg::CPU, Parameters, Real> {
+public:
+  using this_type = CtintWalkerSubmatrix<linalg::GPU, Parameters, Real>;
+  using BaseClass = CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>;
+  using RootClass = CtintWalkerBase<Parameters, Real>;
+
+  using typename BaseClass::Data;
+  using typename BaseClass::Profiler;
+  using typename BaseClass::Rng;
+  using typename BaseClass::CudaStream;
+
+  CtintWalkerSubmatrix(const Parameters& pars_ref, const Data& /*data_ref*/, Rng& rng_ref,
+                       int id = 0);
+
+  void computeM(std::array<dca::linalg::Matrix<Real, linalg::GPU>, 2>& m_accum);
+
+  void initialize();
+
+  void doSweep() override;
+
+  void synchronize();
+
+  using BaseClass::order;
+  using RootClass::get_stream;
+
+  std::size_t deviceFingerprint() const {
+    std::size_t res = 0;
+    for (int s = 0; s < 2; ++s) {
+      res += M_dev_[s].deviceFingerprint();
+      res += Gamma_inv_dev_[s].deviceFingerprint();
+      res += D_dev_[s].deviceFingerprint();
+      res += G_dev_[s].deviceFingerprint();
+      res += G0_dev_[s].deviceFingerprint();
+    }
+    return res;
+  }
+
+private:
+  void doStep() override;
+
+  void computeMInit();
+  void computeGInit();
+  void updateM();
+
+  void uploadConfiguration();
+
+protected:
+  // For testing purposes:
+  void doStep(int n_moves_to_delay);
+
+protected:
+  using BaseClass::configuration_;
+  using BaseClass::M_;
+
+private:
+  template <linalg::DeviceType dev = linalg::GPU>
+  using MatrixView = linalg::MatrixView<Real, dev>;
+  using Matrix = linalg::Matrix<Real, linalg::CPU>;
+
+  DeviceConfigurationManager device_config_;
+
+  using BaseClass::removal_list_;
```

Comment:
reordered declarations in 2e6a34631ba8f3aa710b6b00e4b3968b11117b94. I don't understand which comments you are referring to. Do you mean using BaseClass::foo_ inside a method, rather than bringing the member into the class scope?